### DISCUSSION TITLE
Optional properties don't work with timestamps

### DISCUSSION THREAD

**Discussion by alexrififi** [spatie/laravel-data#188] [Q&A]
Greetings gentlemen!

```php
class SongData extends Data
{
    public function __construct(
        public string $title,
        public Carbon|Optional $created_at,
    ) {
    }
}
```

and

```php
$song = Song::select('title')->first(); // or Song::find(1, 'title')
$data = SongData::from($song);
```

gets the following error

> SongData::__construct(): Argument #2 ($created_at) must be of type Illuminate\Support\Carbon|Spatie\LaravelData\Optional, null given, called in /var/www/html/vendor/spatie/laravel-data/src/Resolvers/DataFromArrayResolver.php on line 57

## Reason

ModelNormalizer forcibly adds date attributes as Data properties

https://github.com/spatie/laravel-data/blob/12a91557e9a5596d48be9b5aaf61eae1c1806fcb/src/Normalizers/ModelNormalizer.php#L18-L26

----

What to do when I wanna timestamps to be optional?

**Comment by rubenvanassche** [comment]
Shouldn't this:

```php
class SongData extends Data
{
    public function __construct(
        public string $title,
        public Carbon|Optional|null $created_at,
    ) {
    }
}
```

Since you're providing a value for `created_at` because it exists in your model, the property is technically not Optional and should be nullable.

**Reply by alexrififi** [reply]
Model not providing a value for `created_at`

```php
Song::find(1, 'title')->toArray();

//        [
//            "title" => "Never Gonna Give You Up"
//        ]
```

I expect in the Data also be missing the `created_at`, but it is forcibly added

**Reply by rubenvanassche** [reply]
`created_at` is added becaus it is a part of your model. You should disable timestamps on your model if you don't have any by setting the `$timestamps = false` property.
